### PR TITLE
Provide options.scriptExtensions for 2.x

### PR DESCRIPTION
**Pull request by supasate**
### Description
Provide scriptExtensions option for custom loading.
Due to require.extensions is not used by some framework (e.g. jest). So, this feature allows specifying file extensions supported by custom loaders.

#### Related issues
strongloop/loopback#3204

### Checklist

- [x] New tests added or existing tests modified to cover all changes
- [x] Code conforms with the [style
  guide](http://loopback.io/doc/en/contrib/style-guide.html)

### CODE REVIEW

**Comment by slnode** [comment]
Can one of the admins verify this patch? To accept patch and trigger a build add comment ".*ok\W+to\W+test.*"

**Comment by slnode** [comment]
Can one of the admins verify this patch?

**Comment by slnode** [comment]
Can one of the admins verify this patch?

**Comment by slnode** [comment]
Can one of the admins verify this patch?

**Comment by bajtos** [comment]
@supasate thank you for the pull request! Does @clarkorz's https://github.com/strongloop/loopback-boot/pull/240#discussion_r107839885 apply here too?

> The third param `onlyScriptsExportingFunction` cause `.js` files of models will not be loaded. 
> Line 142 in utils.js checking `typeof require.extensions['.js'] === 'function'` which will be false when using jest.

**Comment by bajtos** [comment]
@slnode  ok to test

File: lib/compiler.js [hunk 1]
```diff
@@ -38,6 +38,8 @@ module.exports = function compile(options) {
 
   var appRootDir = options.appRootDir = options.appRootDir || process.cwd();
   var env = options.env || process.env.NODE_ENV || 'development';
+  var scriptExtensions = options.scriptExtensions = options.scriptExtensions ||
```
Comment (bajtos) [inline lib/compiler.js]: Modifying input arguments (setting `options.scriptExtensions`) is a bad practice, we have been bitten by this in the past. Please treat `options` are readonly/frozen.

I see that you are accessing `options.scriptExtensions` at https://github.com/strongloop/loopback-boot/pull/242/files#diff-ee97c5091b89979aace94674818996baR722.

Here are two of many possible solutions how to keep `options` frozen and still have `scriptExtensions` set:

```js
// use prototypal inheritance
options = Object.create(options);

// shallow-copy properties
options = util._extend({}, options);
```

(A line like the above would go to L38-39 above, before we start processing the `options` object.)

File: lib/compiler.js [hunk 2]
```diff
@@ -204,9 +209,10 @@ function tryReadDir() {
 }
 
 function buildAllModelInstructions(rootDir, modelsConfig, sources,
-    modelDefinitions) {
-  var registry = verifyModelDefinitions(rootDir, modelDefinitions) ||
-    findModelDefinitions(rootDir, sources);
+    modelDefinitions, scriptExtensions) {
+  var registry = verifyModelDefinitions(rootDir, modelDefinitions,
+                  scriptExtensions) ||
+                  findModelDefinitions(rootDir, sources, scriptExtensions);
```
Comment (bajtos) [inline lib/compiler.js]: The indentation makes it difficult to distinguish function parameters from the second function call. I am proposing the following style:

```js
  var registry = verifyModelDefinitions(rootDir, modelDefinitions,
                                        scriptExtensions);

  if (!registry) {
    registry = findModelDefinitions(rootDir, sources, scriptExtensions);
  }
```

File: lib/compiler.js [hunk 3]
```diff
@@ -38,6 +38,8 @@ module.exports = function compile(options) {
 
   var appRootDir = options.appRootDir = options.appRootDir || process.cwd();
   var env = options.env || process.env.NODE_ENV || 'development';
+  var scriptExtensions = options.scriptExtensions = options.scriptExtensions ||
+                                                    _.keys(require.extensions);
```
Comment (bajtos) [inline lib/compiler.js]: Considering that you are converting this  array back into an object down in `isPreferredExtension`, I think it's better to make the conversion right here.

```js
var scriptExtensions = options.scriptExtensions ?
  arrayToObject(scriptExtensions) : 
  require.extensions;
```

File: lib/compiler.js [hunk 4]
```diff
@@ -644,19 +657,28 @@ function getExcludedExtensions() {
...
+function arrayToObject(array) {
+  return array.reduce(function(obj, val) {
+    obj[val] = val;
+    return obj;
+  }, {});
+}
+
+function isPreferredExtension(filename, extensions) {
+  var includeExtensions = extensions ?
+    arrayToObject(extensions) : require.extensions;
```
Comment (bajtos) [inline lib/compiler.js]: Since `isPreferredExtension` is a private method, there is no need to keep support for the old one-arg API.

```js
function isPreferredExtension(filename, includeExtensions) {
  assert(!!includeExtensions, '"includeExtensions" argument is required');
  // the code below remains unchanged
  var ext = path.extname(filename);
  return (ext in includeExtensions) && !(ext in getExcludedExtensions());}
```

File: test/fixtures/simple-app/boot/custom.customjs [hunk 5]
```diff
@@ -0,0 +1,8 @@
+process.bootFlags.push('customLoaded');
+module.exports = function(app, callback) {
+  process.bootFlags.push('customStarted');
+  process.nextTick(function() {
+    process.bootFlags.push('customFinished');
+    callback();
+  });
```
Comment (bajtos) [inline test/fixtures/simple-app/boot/custom.customjs]: This is unnecessarily complex. The thing to verify in your test is whether loopback-boot recognises a custom extension. There is no need to verify how boot script are loaded/executed again.

```js
'use strict';

module.exports = function(app, callback) {
  process.bootFlags.push('customjs');
  callback();
};
```

Same comment applies to `custom.customjs2`.

File: test/executor.test.js [hunk 6]
```diff
@@ -299,6 +299,27 @@ describe('executor', function() {
       });
     });
 
+    it('should load scripts specified in options.scriptExtensions',
```
Comment (bajtos) [inline test/executor.test.js]:
```js
it('searches boot file extensions specified in options.scriptExtensions')
```

See also http://loopback.io/doc/en/contrib/style-guide.html#test-naming

**Comment by bajtos** [comment]
@supasate also please take a look at https://github.com/strongloop/loopback-boot/pull/240#discussion_r107839885 - please add a test reproducing the issue (I think the trick will be to use a model file with a custom extension that's not registered in `required.extensions`.)

**Comment by bajtos** [comment]
Also please rebase your patch on top of the latest `2.x` branch, I fixed the failing Node 0.10 builds in #244.

**Comment by supasate** [comment]
Hi @bajtos, I've rebased and fixed issues based on your comments. I made a sample repo at https://github.com/supasate/loopback-boot-with-jest. I'll squash later if it's alright.

**Comment by supasate** [comment]
I also found another issue about file name resolving (https://github.com/strongloop/loopback/issues/2201) (not related to this PR but it will make `jest` failed to run).

File: lib/compiler.js [hunk 7]
```diff
@@ -670,51 +699,52 @@ function fixFileExtension(filepath, files, onlyScriptsExportingFunction) {
...
+          //results.push(otherFile);
+        //}
       }
     }
   });
   return (results.length > 0 ? results[0] : undefined);
 }
 
-function resolveAppScriptPath(rootDir, relativePath, resolveOptions) {
+function resolveAppScriptPath(rootDir, relativePath, resolveOptions,
```
Comment (bajtos) [inline lib/compiler.js]: Please keep `resolveOptions` as the last function arg. I would consider converting `scriptExtensions` into a property in `resolveOptions`, to keep the number of function arguments reasonably short.

File: lib/compiler.js [hunk 8]
```diff
@@ -670,51 +699,52 @@ function fixFileExtension(filepath, files, onlyScriptsExportingFunction) {
...
-function buildAllMixinInstructions(appRootDir, mixinDirs, mixinSources, options,
-  modelInstructions) {
-  var extensions = _.without(_.keys(require.extensions),
-    _.keys(getExcludedExtensions()));
-
+function buildAllMixinInstructions(appRootDir, mixins, mixinDirs, mixinSources,
+                                   scriptExtensions, normalization,
+                                   modelInstructions) {
   // load mixins from `options.mixins`
-  var sourceFiles = options.mixins || [];
```
Comment (bajtos) [inline lib/compiler.js]: What is the rationale for this change? I mean why to pass `options` properties as individual arguments? `buildAllMixinInstructions` is already accepting way too many arguments, we should look into ways how to make that list shorter, not longer...

File: lib/compiler.js [hunk 9]
```diff
@@ -670,51 +699,52 @@ function fixFileExtension(filepath, files, onlyScriptsExportingFunction) {
 
       if (!(otherFileExtension in getExcludedExtensions()) &&
         path.basename(f, otherFileExtension) == basename) {
-        if (!onlyScriptsExportingFunction)
-          results.push(otherFile);
-        else if (onlyScriptsExportingFunction &&
-          (typeof require.extensions[otherFileExtension]) === 'function') {
-          results.push(otherFile);
-        }
+        // if (!onlyScriptsExportingFunction)
```
Comment (bajtos) [inline lib/compiler.js]: We don't keep commented-out code around, please remove.

If I understand this code correctly, the intention was:

 - return a file with any extension when `onlyScriptsExportingFunction` is not set
 - return a file with an extension that can be loaded by Node.js runtime (an extension that is registered in `require.extensions`) when `onlyScriptsExportingFunction` is set

We should preserve these two modes.

I am proposing to replace `onlyScriptsExportingFunction` arg with `scriptExtensions` and modify all function callers accordingly.

```js
// implementation here
if (!scriptExtensions || otherFileExtension in scriptExtensions)
  results.push(otherFile);

// usage - flag not set
fixFileExtension(filepath, files)

// usage -flag set
fixFileExtension(filepath, files, scriptExtensions)
```

**Comment by bajtos** [comment]
@supasate thank you for the pull request, please take a look at the comment I left above. Also please rebase your patch on top of the latest `2.x` branch.

File: lib/compiler.js (hunk 9 above)
Comment (supasate) [inline lib/compiler.js]: Oops! That's my mistake. I tested something and forgot to uncomment. I'll fix as you suggested.

File: lib/compiler.js (hunk 8 above)
Comment (supasate) [inline lib/compiler.js]: I tried to make code look consistent with [how `mixinDirs` and `mixinSources` are declared  and passed to the function.](https://github.com/supasate/loopback-boot/blob/8fef5f6fa8b0d2e37b1c38877805fbfeb656dbd6/lib/compiler.js#L112-L113).

Do you want to refactor those arguments by passing only `options` and set their default values in `buildAllMixinInstructions` instead?

**Comment by supasate** [comment]
About rebasing on top of the latest `2.x` branch, do you mean latest of `2.x` branch or `2.x-latest` branch? (I've rebased with `2.x` branch because it seems its last commit is more up-to-date than `2.x-latest`.)

File: lib/compiler.js (hunk 7 above)
Comment (supasate) [inline lib/compiler.js]: After refactoring, there is no need to use `scriptExtensions` in this function anymore.

File: lib/compiler.js (hunk 8 above)
Comment (supasate) [inline lib/compiler.js]: I refactored by passing only `options`.

**Comment by bajtos** [comment]
Landed, thank you for the contribution!

**Comment by bajtos** [comment]
Released in `loopback-boot@2.24.0` 🎉

**Comment by supasate** [comment]
Thanks @bajtos ! 

I think you can review the 3.x branch also. The code is almost the same except that it is splitted into several files.
